### ISSUE TITLE
Weird message when doing panzerschreck

### ISSUE THREAD

**Issue by x3003** [Realm667/WolfenDoom#512]
When you do the panzerschreck on the training course, there's a weird message at the bottom.

![Screenshot_Doom_20210424_151836](https://user-images.githubusercontent.com/77861029/115960372-4ffdeb00-a511-11eb-9ebe-8af20e422116.png)

**Comment by Tormentor667** [comment]
I guess something is wrong with the changes that have been made to the messages and key displays @AFADoomer

**Comment by Username-N00b-is-not-available** [comment]
This does not seem to be a bug. Do you have a key bound to weapon reloading, @x3003?

**Comment by Tormentor667** [comment]
It definitely is a bug, look here (intro video, the key is displayed right to the message):
<img width="953" alt="2021-04-24_153404" src="https://user-images.githubusercontent.com/10298439/115960591-9142ca80-a512-11eb-8c26-9fb6082dd01b.png">

**Comment by Tormentor667** [comment]
I think there is one space too much between the key and the explanation.
<img width="953" alt="2021-04-24_153404" src="https://user-images.githubusercontent.com/10298439/115960638-e1ba2800-a512-11eb-9891-5f8d98fdc594.png">

**Comment by Username-N00b-is-not-available** [comment]
???
@Tormentor667 How is it a bug? What would you consider proper behaviour in showing these messages?

**Comment by x3003** [comment]
Because you can use the space key as reload, like when you reload the grenates in the training course.

**Comment by AFADoomer** [comment]
The message shows up because you don't have a key bound directly to throw a grenade.  I can add logic to replace that with a "bind a key" message, but that will require string translation.

Plus, grenades can be thrown from the inventory bar with no key bound (which is what I do), so the message probably should be tweaked anyway.

**Comment by x3003** [comment]
Well, normal reloading happens w/ mouse1, so why should I have another key for that ?

**Comment by AFADoomer** [comment]
Is it bound only as "+rebind", or something else also?  My mouse binds (mouse2 for use, mouse1 for attack) show up fine.

**Comment by Username-N00b-is-not-available** [comment]
Sometimes you would like to reload after shooting half of the magazine, as it is rarely a good idea to come into a room full of enemies with a half-loaded gun.
This is of course a matter of personal preference.

**Comment by x3003** [comment]
Yes. I never used the 'reload' option.

**Comment by Username-N00b-is-not-available** [comment]
Well, since the reload key is bound by default in BoA, and you have to unbind it to get this behaviour, I would like to close this issue. It would be the player's own fault if they unbind some key and then become very surprised that something minor like this shows up unexpectedly. "Сам себе злобный Буратино", as we Russians sometimes say.
By the way, the displayed text has a meaning, ??? obviously means that the keybind is not present, and +reload is itself the console command equivalent to pressing the reload key. The keys are bound to such commands: when you press W (or the up arrow, or RMB, whatever your bind for forward moving is), the console receives a ```+forward``` command, and when you unpress it, a ```-forward``` command is sent. These commands are processed by the engine, and your player moves.

**Comment by DoomJedi** [comment]
I think I have to disagree here, or at least add a remark:

People don't have to actively "unbind" Reloading (or any other action for that matter), all they need it to bind R for other action, because they are used to use R for it in other games, or for whatever reasons. It's hard to remember all already used bindings, so this can happen without intent and see no need to punish player for it - for mere trying to bind different action to R letter.

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi Good point on not punishing the players, but this is not punishing in any way. Instead, this particular 'strange' message _reminds_ the player to bind the action to some key ("Why did this message appear instead of a regular key bind?")

**Comment by AFADoomer** [comment]
I'm working on cleaning this up so that it displays a "Bind a key: " message instead of the ???.

**Comment by Tormentor667** [comment]
> ???
> @Tormentor667 How is it a bug? What would you consider proper behaviour in showing these messages?

Now I understood - this is the „hint“ you were all working on. Well, as of now it looks strange how it is displayed - behind the sentence instead of replacing the string inline. Maybe this needs a different look to be different from the regular displayed hint?

**Comment by AFADoomer** [comment]
How does it look for everyone after 1668d00345f8d0c2e8b70ac256cb2cc5521e9b00?

You now get a message telling you to bind a key if the action is not bound, and if the action starts with "use", you "or <activate item>" is appended automatically.

The "Bind a key:" message and "or" need translations: BINDKEY and WORD_OR (I set "or" to fall back to 'or' to match the engine's default presentation of key bind info in ACS).

**Comment by Talon1024** [comment]
I removed the binding for primary fire, and I saw <Bind a key: +attack> at the bottom of the screen when I started the training course.

If I didn't know how GZDoom worked, I would be very confused if I saw that. Like, what the heck is `+attack` supposed to mean?

**Comment by AFADoomer** [comment]
Yeah, I think I will add a lookup of the action bind name if it exists.

Non-English languages already get 'or' right now because it's hard-coded into the engine.

**Comment by AFADoomer** [comment]
Updated again in f77a0750a687acdc158c6cac1932b6442fa9c958.

**Comment by x3003** [comment]
I hope everyone knows what bind a key is, but you can also simply use 'space' to reload some more.

**Comment by Username-N00b-is-not-available** [comment]
@x3003 Do you mean here and in https://github.com/Realm667/WolfenDoom/issues/512#issuecomment-826105373 pressing "use" (Space) on the shelves? We are talking about reloading weapon magazines, i. e. putting there ammo which you _already_ own.
When you interact with the shelves, you do not 'reload' any weapon, you simply take more grenades from an external storage. It could be called 'restacking' on grenades, but not reloading.
***
Related: how difficult would it be to add a hint to the Customize Controls menu which would pop up when a previous keybind jumps off (for example, when you bind something to R after a fresh install, and there is no Reload key anymore)? I am thinking about a message shown at the center of the screen, as a dialogue box which needs confirmation:

```
<R> is already bound to <Reload>.
Bind <Throw Grenade> to <R> (y/n)?
```

or just a hint message:

```
Attention: <R> is no longer bound to <Reload>.
```

**Comment by x3003** [comment]
Well, on c1m4 there's a message to use the 'use' key **and** to do some binding. This is confusing.

![Screenshot_Doom_20210425_142752](https://user-images.githubusercontent.com/77861029/115993495-49d04300-a5d3-11eb-81c1-27682a2fb844.png)

**Comment by AFADoomer** [comment]
> Related: how difficult would it be to add a hint to the Customize Controls menu which would pop up when a previous keybind jumps off (for example, when you bind something to R after a fresh install, and there is no Reload key anymore)? I am thinking about a message shown at the center of the screen, as a dialogue box which needs confirmation:
> 
> ```
> <R> is already bound to <Reload>.
> Bind <Throw Grenade> to <R> (y/n)?
> ```
> 
> or just a hint message:
> 
> ```
> Attention: <R> is no longer bound to <Reload>.
> ```

Do-able I think, but probably not for release.

**Comment by Tormentor667** [comment]
I think we should collect things for a 3.1 release in a seperate issue so we won't forget about all the good ideas.

**Comment by Username-N00b-is-not-available** [comment]
@x3003 What do you think, does the hint message get shown properly now?

**Comment by x3003** [comment]
Well, on the panzerschreck it still says bind key: weapon reload. If there's no key binding, why not leave this msg empty ?
This key isn't assigned to a new gzdoom profile I think, so any new user who is gonna download boa in a while, doesn't know about it unless they have played it before.
The other msgs look fine.

**Comment by AFADoomer** [comment]
Because just because you choose to not use the reload key doesn't mean that everyone will.

So everything is working as intended and this can be closed.

**Comment by Username-N00b-is-not-available** [comment]
The reload/quick kick/throw grenade/crouch (not toggle crouching) keys are definitely not bound by default, you can see this by deleting the configuration file.
Where are the default binds being assigned? I thought of [DEFBINDS](https://zdoom.org/wiki/DEFBINDS), but there is no such file in Blade of Agony, so maybe something else is used.

**Comment by Tormentor667** [comment]
> The reload/quick kick/throw grenade/crouch (not toggle crouching) keys are definitely not bound by default, you can see this by deleting the configuration file.

These should be bound by default. Isn’t there a Keyconfig in the GAMEINFO @Ozymandias81 ?

**Comment by x3003** [comment]
>Because just because you choose to not use the reload key doesn't mean that everyone will.

Yeah, but the reload button is a *forced* reload, which you don't need there, maybe in the levels, when the clip is not fully empty yet, but you don't need it with the panzerschreck which can fire only 1 anyway.

**Comment by DoomJedi** [comment]
We need to assign those actions by default

**Comment by Ozymandias81** [comment]
Typing them down now.

**Comment by Username-N00b-is-not-available** [comment]
It seems that everything here has been resolved.

**Comment by Username-N00b-is-not-available** [comment]
@Ozymandias81 @Tormentor667 It sadly turns out that DEFBINDS won't work properly with r195: https://forum.zdoom.org/viewtopic.php?f=7&t=68292 (and they do not work for me now). But choosing a stable version for 3.1 will take care of that.

**Comment by Tormentor667** [comment]
I think we should wait for the next stable GZDoom release anyway before releasing 3.1

**Comment by Ozymandias81** [comment]
Yes as I suggested, we'll have time until next stable release. I guess it will happen for the end of next month.

**Comment by Blue-Shadow** [comment]
> 
> 
> It sadly turns out that DEFBINDS won't work properly with r195: https://forum.zdoom.org/viewtopic.php?f=7&t=68292

That's an old issue which was fixed in GZDoom 4.5.0 (the thread is from last year, not this one).

**Comment by Username-N00b-is-not-available** [comment]
@Blue-Shadow Ah, sorry... It did not work for me though for some reason today when I deleted the config files (and it does not work now). Just saw April dates and thought it was related to my problem.
